crates/ty_python_semantic/src/types/infer.rs

```diff
@@ -6438,40 +6438,31 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
                 // If the legacy typevar is still unbound after that search, and we are in a
                 // context that binds unbound legacy typevars (i.e., the signature of a generic
                 // function), bind it with that context.
-                let find_legacy_typevar_binding = |typevar: TypeVarInstance<'db>| {
-                    enclosing_generic_contexts(
+                Type::TypeVar(typevar) if typevar.is_legacy(self.db()) => bind_legacy_typevar(
+                    self.db(),
+                    self.context.module(),
+                    self.index,
+                    self.scope().file_scope_id(self.db()),
+                    self.legacy_typevar_binding_context,
+                    typevar,
+                )
+                .map(Type::TypeVar)
+                .unwrap_or(ty),
+                Type::KnownInstance(KnownInstanceType::TypeVar(typevar))
+                    if typevar.is_legacy(self.db()) =>
+                {
+                    bind_legacy_typevar(
```

Comment:
I think this is kind of orthogonal to this PR, and relates instead to a previous PR that landed last week while I was out -- but it doesn't feel right to me that we are handling `Type::TypeVar(...)` and `Type::KnownInstance(KnownInstanceType::TypeVar(...))` in such a parallel way here. It seems like we are missing some invariants that we ought to be enforcing? I think it should be an invariant that `Type::TypeVar` can never possibly wrap an unbound `TypeVarInstance` in the first place, so a `Type::TypeVar` should never need re-binding. `Type::TypeVar` represents an actual value of the variable type -- this is meaningless for an unbound typevar.

Reply:
Makes sense. That issue proposes one possible fix (split `TypeVarInstance` into separate Rust types for bound and unbound type variables) which would clearly allow us to enforce the invariant, but even short of that it seems like we could have some asserts/unreachables in place of handling that should be unnecessary (like here, I don't think we should ever call `bind_legacy_typevar` on something inside `Type::TypeVar`, that should always be already bound) to better clarify the semantics. Doesn't have to be in this PR, though.